library/bignum.c

```diff
@@ -1373,49 +1374,43 @@ int mbedtls_mpi_sub_int( mbedtls_mpi *X, const mbedtls_mpi *A, mbedtls_mpi_sint
  *
  * Add \p b * \p s to \p d.
  *
- * \param i             The number of limbs of \p s.
+ * \param[in,out] d     The bignum to add to.
+ * \param d_len         The number of limbs of \p d. This must be
+ *                      at least \p s_len.
```

Comment:
@mpg does Mbed TLS have a mechanism for `assert()`ing requirements like this? or is it considered not worth it, even from a documentation-in-code perspective?

Reply:
From all I know, Mbed TLS does not yet have such a mechanism, though it would arguably be useful for debug builds.

Reply:
No, we don't `assert()`, for several reasons. One reason is that the code must be able to run in environments that don't have `assert()`, and we don't want to maintain yet another compilation option `HAVE_ASSERT`. Another is that we want to zeroize confidential data even if (especially if!) something goes wrong. So if a requirement is violated, the only recourse is to return an error status. But that is not possible in low-level bignum functions that return void.

So far our general approach in bignum is to have public functions that check requirements and return error codes (also those functions may resize objects), and low-level functions that don't check requirements and return void (and those functions may not resize objects).

Reply:
Another reason not to use `assert` is that it's typically disabled in production, which is where it's the most useful. If you want a debug-only assert, you can have extra code under `MBEDTLS_TEST_HOOKS` that should not change the execution of the function, but can mark the test as failed.